File: linked_list/cc/linked_list.cc
#include "linked_list.h"

namespace mf
{

template <class T>
LinkedList<T>::~LinkedList()
{
  LinkedListNode<T> *current = head_;
  LinkedListNode<T> *next = nullptr;

  while (current) {
    next = current->get_next();
    delete current;
    current = next;
  }
}

template <class T>
int LinkedList<T>::size() const
{
  int count = 0;
  for (LinkedListNode<T> *i = head_; i != nullptr; i = i->get_next()) {
    count++;
  }
  return count;
}

template <class T>
bool LinkedList<T>::is_empty() const
{
  return head_ == nullptr;
}

template <class T>
std::optional<T> LinkedList<T>::value_at(int index) const
{
  if (index < 0 || index >= this->size())
    return std::nullopt;

  LinkedListNode<T> *node = head_;
  for (int i = 0; i != index; i++, node = node->get_next());

  return node->get_value();
}

template <class T>
void LinkedList<T>::push_front(T value)
{
  LinkedListNode<T> *node = new LinkedListNode<T>(value);
  node->set_prev(nullptr);
  node->set_next(head_);

  if (head_ == nullptr)
    tail_ = node;
  else
    head_->set_prev(node);

  head_ = node;
}

template <class T>
std::optional<const T> LinkedList<T>::pop_front()
{
  if (head_ == nullptr)
    return std::nullopt;

  LinkedListNode<T> *node = head_;
  T value = node->get_value();

  head_ = head_->get_next();
  if (head_ == nullptr)
    tail_ = nullptr;
  else
    head_->set_prev(nullptr);

  delete node;
  return value;
}

template <class T>
void LinkedList<T>::push_back(T value)
{
  LinkedListNode<T> *node = new LinkedListNode<T>(value);
  node->set_prev(tail_);
  node->set_next(nullptr);

  if (head_ == nullptr)
    head_ = node;
  else
    tail_->set_next(node);

  tail_ = node;
}

template <class T>
std::optional<const T> LinkedList<T>::pop_back()
{
  if (head_ == nullptr)
    return std::nullopt;

  LinkedListNode<T> *node = tail_;
  T value = node->get_value();

  tail_ = tail_->get_prev();
  if (tail_ == nullptr)
    head_ = nullptr;
  else
    tail_->set_next(nullptr);

  delete node;
  return value;
}

template <class T>
std::optional<const T> LinkedList<T>::front() const
{
  if (head_ == nullptr)
    return std::nullopt;
  return head_->get_value();
}

template <class T>
std::optional<const T> LinkedList<T>::back() const
{
  if (head_ == nullptr)
    return std::nullopt;
  return tail_->get_value();
}

template <class T>
void LinkedList<T>::insert(int index, T value)
{
  if (index < 0 || index >= this->size())
    return;

  LinkedListNode<T> *node = new LinkedListNode<T>(value);
  LinkedListNode<T> *in = head_;
  for (int i = 0; i != index; i++, in = in->get_next());

  node->set_next(in);
  node->set_prev(in->get_prev());

  if (in->get_prev() == nullptr)
    head_ = node;
  else
    in->get_prev()->set_next(node);

  in->set_prev(node);
}

template <class T>
void LinkedList<T>::erase(int index)
{
  if (index < 0 || index >= this->size())
    return;
  
  LinkedListNode<T> *in = head_;
  for (int i = 0; i != index; i++, in = in->get_next());

  if (in->get_prev() == nullptr)
    head_ = in->get_next();
  else
    in->get_prev()->set_next(in->get_next());

  if (in->get_next() == nullptr)
    tail_ = in->get_prev();
  else
    in->get_next()->set_prev(in->get_prev());

  delete in;
}

template <class T>
void LinkedList<T>::reverse()
{
  if (head_ == nullptr || head_->get_next() == nullptr)
    return;

  tail_ = head_;
  LinkedListNode<T> *tmp = nullptr;

  for (LinkedListNode<T> *i = head_; i != nullptr; i = i->get_prev()) {
    tmp = i->get_prev();
    i->set_prev(i->get_next());
    i->set_next(tmp);
  }

  head_ = tmp->get_prev();
}

}  // namespace mf

File: data-structures/linked_list/python/linked_list.py
#!/usr/bin/evn python2


class LinkedListNode(object):
  def __init__(self, value):
    self._value = value
    self._prev = None
    self._next = None

  @property
  def value(self):
    return self._value

  @value.setter
  def value(self, value):
    self._value = value

  @property
  def next(self):
    return self._next

  @next.setter
  def next(self, next):
    self._next = next

  @property
  def prev(self):
    return self._prev

  @prev.setter
  def prev(self, prev):
    self._prev = prev


class LinkedList(object):
  def __init__(self):
    self._head = None  # Instance of LinkedListNode
    self._tail = None  # Instance of LinkedListNode

  def __len__(self):
    size = 0
    node = self._head
    while node is not None:
      size += 1
      node = node.next
    return size

  def is_empty(self):
    return self._head is None

  def push_front(self, value):
    node = LinkedListNode(value)
    node.prev = None
    node.next = self._head

    if self._head is None:
      self._tail = None
    else:
      self._head.prev = node

    self._head = node

  def pop_front(self):
    if self._head is None:
      return

    node = self._head
    value = node.value

    self._head = self._head.next
    if self._head is None:
      self._tail = None
    else:
      self._head.prev = None

    del node
    return value

  def push_back(self, value):
    node = LinkedListNode(value)
    node.prev = self._tail
    node.next = None

    if self._head is None:
      self._head = node
    else:
      self._tail.next = node

    self._tail = node

  def pop_back(self):
    if self._head is None:
      return

    node = self._tail
    value = node.value

    self._tail = self._tail.prev
    if self._tail is None:
      self._head = None
    else:
      self._tail.next = None

    del node
    return value

  def front(self):
    if self._head is None:
      return
    return self._head.value

  def back(self):
    if self._head is None:
      return
    return self._tail.value

  def insert(self, index, value):
    if index < 0 or index >= len(self):
      return

    node = LinkedListNode(value)
    it = self._head
    i = 0
    while i != index:
      it = it.next
      i += 1

    node.next = it
    node.prev = it.prev

    if it.prev is None:
      self._head = node
    else:
      it.prev.next = node

    it.prev = node

  def erase(self, index):
    if index < 0 or index >= len(self):
      return

    it = self._head
    i = 0
    while i != index:
      it = it.next
      i += 1

    if it.prev is None:
      self._head = it.next
    else:
      it.prev.next = it.next

    if it.next is None:
      self._tail = it.prev
    else:
      it.next.prev = it.prev

    del it

  def reverse(self):
    if self._head is None or self._head.next is None:
      return

    self._tail = self._head
    tmp = None
    i = self._head

    while i is not None:
      tmp = i.prev
      i.prev, i.next = i.next, i.prev
      i = i.prev

    self._head = tmp.prev
